### ISSUE TITLE
From scratch implementation

### ISSUE THREAD

**Issue by olivkoch** [SamsungSAILMontreal/TinyRecursiveModels#26]
Is anyone interested in discussing a from-scratch implementation of TRM? I've started doing this on my own and have a few questions regarding discrepencies between the [paper](https://arxiv.org/pdf/2510.04871v1) and the code.

- I can't find the N_supervision loop in the code

- I don't see mentions of the carry structure in the paper. While this is certainly an implementation trick, I'd like to understand what it corresponds to.

Thanks!

**Comment by postmachines** [comment]
Take a closer look at the 'halt_max_steps'. They play the role of N_supervision.

**Comment by olivkoch** [comment]
I am building an implementation of TRM from scratch with the goal to unpack its magic. Contributions are welcome!

https://github.com/olivkoch/nano-trm

**Comment by dah33** [comment]
@aabbas90 so you still truncate the bp to a single supervision step in each case? That removes the path dependency implicit in optimising step one, then using that new model to optimise step two etc.

**Comment by alexander-rakhlin** [comment]
Has anyone noticed that the final checkpoints give mediocre accuracy on ARC2? For example, step_723914 has around 2%, while its predecessor, step_651522, scores 7-8% just like reported in the paper. This applies both to the official checkpoint from HF and to the one I trained myself. I suspect there might be a bug.

**Comment by olivkoch** [comment]
@aabbas90 I do not see a benefit for this change yet but training takes time on my single H100. I will continue iterating on the code and keep you posted.

**Comment by sandorvasas** [comment]
@aabbas90 Hi, did you achieve this 87.9% accuracy on sudoku-extreme or sudoku-extreme-1k? My impl from scratch -using sudoku-extreme-1k- converges to only 40% accuracy.

**Comment by bumi001** [comment]
This is all so that, at all loops, all "slots" in the batch are being used to keep GPU usage high. This does mean that most of the batch's data are never used, but this doesn't matter much here since the program goes over the whole dataset repeatedly.

Does this ensure that every item in the training dataset is used at least once? Is it possible that certain puzzle_ids are never used in training? In cfg_pretrain.yaml, I see, epochs = 100000 and eval_interval: 10000. 

```
    train_epochs_per_iter = config.eval_interval if config.eval_interval is not None else config.epochs
    total_iters = config.epochs // train_epochs_per_iter
```
So, it iterates over the entire dataset 10 times. Suppose, we don't halt early and halt_max_steps is 16. In that case, it would only using every 16th batch. It seems that maybe why halting early produces better results?
